pkg/controller/statefulset/stateful_set_utils.go

```diff
@@ -255,6 +281,124 @@ func newStatefulSetPod(set *apps.StatefulSet, ordinal int) *v1.Pod {
+func newRevision(set *apps.StatefulSet, revision int64) (*apps.ControllerRevision, error) {
+	patch, err := getPatch(set)
+	if err != nil {
+		return nil, err
+	}
+	selector, err := metav1.LabelSelectorAsSelector(set.Spec.Selector)
+	if err != nil {
+		return nil, err
+	}
+	return history.NewControllerRevision(set,
+		controllerKind,
+		selector,
+		runtime.RawExtension{Raw: patch},
+		revision)
+}
+
+// applyRevision returns a new StatefulSet constructed by restoring the state in revision to set. If the returned error
+// is nil, the returned StatefulSet is valid.
+func applyRevision(set *apps.StatefulSet, revision *apps.ControllerRevision) (*apps.StatefulSet, error) {
+	obj, err := scheme.Scheme.DeepCopy(set)
+	if err != nil {
+		return nil, err
+	}
+	clone := obj.(*apps.StatefulSet)
+	patched, err := strategicpatch.StrategicMergePatch([]byte(runtime.EncodeOrDie(patchCodec, clone)), revision.Data.Raw, clone)
+	if err != nil {
+		return nil, err
+	}
+	err = json.Unmarshal(patched, clone)
+	if err != nil {
+		return nil, err
+	}
+	return clone, nil
+}
+
+// nextRevision finds the next valid revision number based on revisions. If the length of revisions
+// is 0 this is 1. Otherwise, it is 1 greater than the largest revision's Revision. This method
+// assumes that revisions has been sorted by Revision.
+func nextRevision(revisions []*apps.ControllerRevision) int64 {
+	count := len(revisions)
+	if count <= 0 {
+		return 1
+	}
+	return revisions[count-1].Revision + 1
+}
+
+// inconsistentStatus returns true if the ObservedGeneration of status is greater than set's
+// Generation or if any of the status's fields do not match those of set's status.
+func inconsistentStatus(set *apps.StatefulSet, status *apps.StatefulSetStatus) bool {
+	return set.Status.ObservedGeneration == nil ||
+		*status.ObservedGeneration > *set.Status.ObservedGeneration ||
+		status.Replicas != set.Status.Replicas ||
+		status.CurrentReplicas != set.Status.CurrentReplicas ||
+		status.ReadyReplicas != set.Status.ReadyReplicas ||
+		status.UpdatedReplicas != set.Status.UpdatedReplicas ||
+		status.CurrentRevision != set.Status.CurrentRevision ||
+		status.UpdateRevision != set.Status.UpdateRevision
+}
+
+// completeRollingUpdate completes a rolling update when all of set's replica Pods have been updated
+// to the updateRevision. status's currentRevision is set to updateRevision and its' updateRevision
+// is set to the empty string. status's currentReplicas is set to updateReplicas and its updateReplicas
+// are set to 0.
+func completeRollingUpdate(set *apps.StatefulSet, status *apps.StatefulSetStatus) {
+	if set.Spec.UpdateStrategy.Type == apps.RollingUpdateStatefulSetStrategyType &&
+		status.UpdatedReplicas == status.Replicas &&
+		status.ReadyReplicas == status.Replicas {
+		status.CurrentReplicas = status.UpdatedReplicas
+		status.CurrentRevision = status.UpdateRevision
+		status.UpdateRevision = ""
+		status.UpdatedReplicas = 0
```

Comment:
It seems counter intuitive to me that when an update completes successfully that the reported current version of the deployed product would be inconsistent with the version newly deployed product. 

Consider the following scenario.

As implemented, when a user deploys a StatefulSet, the user will see all replicas as currentReplicas labeled with the currentRevision. If the user gets the Pods in their set, the version of the software that they initially deployed (say MySQL 5.7.17) will be identical for all of these Pods and correspond to the currentRevision of the StatefulSet as communicated by its status. If the user gets the StatefulSet and all of the Pods in the StatefulSet, they will see that revision labels of the Pods match the currentRevision of the StatefulSet. They can correlate this back to the product they initially deployed.

When the user updates the StatefulSet, say to deploy MySQL 5.7.18, while the update is progress all Pods running 5.7.17 are labeled with currentRevision, all Pods with MySQL 5.7.18 are labeled with updateRevision, and the current and updated replicas provide accounting for the number of Pods with each version. This models the users intention. The current version that they successfully deployed is 5.7.17 and update they wish to deploy 5.7.18. If the user gets the StatefulSet and its Pods, they can clearly distinguish which Pods correspond to which versions by their labeling.

When the update completes, the current version of the software is 5.7.18, the currentRevision of the StatefulSet is set to the revision (previously updateRevision) that corresponds to this software, and the replica counters are updated to reflect the fact that all Pods in the StatefulSet are at this version. This communicates that the current state of the system is at the state that the user desired when applying the update. If the user gets the StatefulSet and all of its generated Pods, they will see that their desired state has been realized, the current revision of the StatefulSet is the one that they applied. All Pods in the StatefulSet are labeled with this revision, and all are running MySQL 5.7.18.

If I understand the approach that you are suggesting, I think that we will send the user false (or at least confusing) signals. Consider that, in the same scenario as above, when user initially deploys the StatefulSet the the currentRevision will correspond to MySQL 5.7.17 as above, and when the update completes the currentRevision will still correspond to MySQL 5.7.17. The updateRevision will correspond to 5.7.18. Even though the update is complete, if the user gets the StatefulSet and all of its Pods, they will see that no Pods correspond to the currentRevision. This seems very counter-intuitive to me. Whenever I update software or configuration (in any context), if the update is successful, I expect the system to report that current version of the software is the version that I applied. If we take this, approach we signal to the user that the current version of the software is the prior version, and the update version is the version that they applied.
Also, if the user updates the StatefulSet again, the currentRevision will switch to the updateRevision at the same time the updateRevision is updated to indicate a new version of the product is being deployed. So we will eventually signal that the currentRevision of the StatefulSet matches the update the user previously applied, but we will only do so when they perform another update. This indicates that the updateRevision only becomes current when a new update is performed.
This approach would mean that the currentRevision is current version during initial deployment, but after the first update, the currentRevision is really the previous version, and the update revision is the currentRevision. The exception to this is when an update is actually in progress. Only then would the current and update revisions accurately model the users intentions.